pgml-cms/blog/fire.md

```diff
@@ -0,0 +1,228 @@
+        "markdown": {
+            "splitter": {"model": "markdown"},
+            "semantic_search": {
+                "model": "mixedbread-ai/mxbai-embed-large-v1",
+            },
+        },
+    },
+)
+collection = Collection("fire-crawl-demo-v0")
+
+# Add our Pipeline to our Collection
+async def add_pipeline():
+    await collection.add_pipeline(pipeline)
+```
+
+This Pipeline configuration tells Korvus how to process our documents. It specifies that we'll be working with markdown content, using a markdown-specific splitter, and the `mixedbread-ai/mxbai-embed-large-v1` model for semantic search embeddings.
+
+See the [Korvus guide to construction Pipelines](https://postgresml.org/docs/open-source/korvus/guides/constructing-pipelines) for more information on Collections and Pipelines.
+
+### Web Crawling with Fire Crawl
+
+The `crawl()` function demonstrates how to use Fire Crawl to scrape a website:
+
+```python
+def crawl():
+    crawl_url = "https://postgresml.org/blog"
+    params = {
+        "crawlerOptions": {
+            "excludes": [],
+            "includes": ["blog/*"],
+            "limit": 250,
+        },
+        "pageOptions": {"onlyMainContent": True},
+    }
+    job = app.crawl_url(crawl_url, params=params, wait_until_done=False)
+    while True:
+        print("Scraping...")
+        status = app.check_crawl_status(job["jobId"])
+        if not status["status"] == "active":
+            break
+        time.sleep(5)
+    return status
+```
+
+This function initiates a crawl of the PostgresML blog, focusing on blog posts and limiting the crawl to 250 pages. It then periodically checks the status of the crawl job until it's complete.
+
+Alternativly to sleeping, we could set the `wait_until_done` parameter to `True` and the `crawl_url` method would block until the data is ready.
+
+
+### Processing and Indexing the Crawled Data
+
+After crawling the website, we need to process and index the data for efficient searching. This is done in the `main()` function:
+
+```python
+async def main():
+    # Add our Pipeline to our Collection
+    await add_pipeline()
+
+    # Crawl the website
+    results = crawl()
+
+    # Construct our documents to upsert
+    documents = [
+        {"id": data["metadata"]["sourceURL"], "markdown": data["markdown"]}
+        for data in results["data"]
+    ]
+
+    # Upsert our documents
+    await collection.upsert_documents(documents)
+```
+
+This code does the following:
+1. Adds the previously defined pipeline to our collection.
+2. Crawls the website using the `crawl()` function.
+3. Constructs a list of documents from the crawled data, using the source URL as the ID and the markdown content as the document text.
+4. Upserts these documents into the collection. The pipeline automatically splits the markdown and generates embeddings for each chunk storing it all in Postgres.
+
+### Performing RAG
+
+With our data indexed, we can now perform RAG:
+
+```python
+async def do_rag(user_query):
```

Comment:
This function name cracks me up. https://en.wikipedia.org/wiki/Durag#Spelling_and_etymology

Reply:
haha I know its hilarious